File: examples/python/render_segmentation.py
#!/usr/bin/env python3
"""Render a gate-segmentation mask from Python.

Python port of examples/render_segmentation.cpp: loads the same three YAML
configs, renders a mask for a hardcoded drone pose, and writes it as a
grayscale PNG.
"""
import argparse

import cv2

from segment_gate_renderer import DronePose, GateRenderer


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--gates-config", default="config/gates_config.yaml")
    parser.add_argument("--drone-config", default="config/config.yaml")
    parser.add_argument("--camera-config", default="config/camera_calibration.yaml")
    parser.add_argument("--output", required=True)
    parser.add_argument("-r", "--rectified", action="store_true")
    return parser.parse_args()


def main() -> None:
    args = parse_args()

    renderer = GateRenderer(args.gates_config, args.drone_config, args.camera_config, args.rectified)

    pose = DronePose(x=19.0, y=2.0, z=0.155, roll=0.0, pitch=0.0, yaw=3.13)
    mask = renderer.render(pose)

    if not cv2.imwrite(args.output, mask):
        raise SystemExit(f"error: failed to write {args.output}")


if __name__ == "__main__":
    main()

File: examples/render_segmentation.cpp
// Render a gate-segmentation image as seen from a given drone pose.
//
// Given a gates layout, a drone pose, and a camera calibration, renders a
// grayscale PNG (0 = background, 255 = gate frame) of what the drone's
// camera would see from that pose. Non-interactive port of
// map_error_net's scripts/generate_segmentation.py (default mode only).
#include <iostream>
#include <string>

#include <opencv2/imgcodecs.hpp>

#include "segment_gate/projection.hpp"
#include "segment_gate/scene.hpp"

namespace {

struct Args {
    std::string gatesConfig = "config/gates_config.yaml";
    std::string droneConfig = "config/config.yaml";
    std::string cameraConfig = "config/camera_calibration.yaml";
    std::string output;
    bool rectified = false;
};

Args parseArgs(int argc, char** argv) {
    Args args;
    for (int i = 1; i < argc; ++i) {
        const std::string arg = argv[i];
        auto nextValue = [&]() -> std::string {
            if (i + 1 >= argc) {
                throw std::runtime_error("missing value for " + arg);
            }
            return argv[++i];
        };

        if (arg == "--gates-config") {
            args.gatesConfig = nextValue();
        } else if (arg == "--drone-config") {
            args.droneConfig = nextValue();
        } else if (arg == "--camera-config") {
            args.cameraConfig = nextValue();
        } else if (arg == "--output") {
            args.output = nextValue();
        } else if (arg == "-r" || arg == "--rectified") {
            args.rectified = true;
        } else {
            throw std::runtime_error("unknown argument: " + arg);
        }
    }
    if (args.output.empty()) {
        throw std::runtime_error("--output is required");
    }
    return args;
}

}  // namespace

int main(int argc, char** argv) {
    using namespace segment_gate;

    Args args;
    try {
        args = parseArgs(argc, argv);
    } catch (const std::exception& e) {
        std::cerr << "error: " << e.what() << "\n";
        return 1;
    }

    const auto gates = loadGatesConfig(args.gatesConfig);
    const auto [dronePos, gateDims] = loadDroneConfig(args.droneConfig);
    const CameraCalibration calib = loadCameraCalibration(args.cameraConfig);

    cv::Mat cameraMatrix = calib.cameraMatrix;
    cv::Mat distCoeffs = calib.distCoeffs;
    bool fisheye = true;
    double thetaMax = 89.0 * CV_PI / 180.0;

    if (args.rectified) {
        cameraMatrix = rectifiedCameraMatrix(cameraMatrix, distCoeffs, calib.imageWidth, calib.imageHeight);
        distCoeffs = cv::Mat::zeros(4, 1, CV_64F);
        thetaMax = rectifiedThetaMax(cameraMatrix, calib.imageWidth, calib.imageHeight);
        fisheye = false;
    }

    const cv::Mat mask = renderPose(gates, gateDims, dronePos, calib.tBaseCam, cameraMatrix, distCoeffs,
                                     calib.imageWidth, calib.imageHeight, fisheye, thetaMax);

    if (!cv::imwrite(args.output, mask)) {
        std::cerr << "error: failed to write " << args.output << "\n";
        return 1;
    }

    return 0;
}
